[Spark] Add support for allow-precision-loss in decimal operations

jinchengchenghh opened the pull request:
Each of the decimal operation functions is registered as two functions such as `add_deny_precision_loss` and `add`.
When allowing precision loss, establishing the result type of an arithmetic operation happens according to Hive behavior and SQL ANSI 2011 specification, i.e. rounding the decimal part of the result if an exact representation is not possible. Otherwise, NULL is returned in those cases, as previously.
When not allowing precision loss, not rounding the decimal part.

For example, 
  | decimal(38, 7) + decimal(10, 0) result type | 1.1232154   + 1| decimal(38, 18) * decimal(38, 18)| 0.1234567891011 * 1234.1
-- | -- | -- | -- | --
allow   precision loss | decimal(38, 6) | 2.123215 | decimal(38, 6) | 152.358023
deny precision   loss | decimal(38, 7) | 2.1232154 | decimal(38, 36) | NULL

```
spark-sql (default)> set spark.sql.decimalOperations.allowPrecisionLoss=true;

spark-sql (default)> select cast(0.1234567891011 as decimal(38, 18)) * cast(1234.1 as decimal(38, 18));
152.358023

spark-sql (default)> set spark.sql.decimalOperations.allowPrecisionLoss=false;

spark-sql (default)> select cast(0.1234567891011 as decimal(38, 18)) * cast(1234.1 as decimal(38, 18));
NULL
```

Spark implementation: https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala#L814

jinchengchenghh commented:
Can you help review this PR? Thanks! @rui-mo @PHILO-HE

jinchengchenghh commented:
hi @rui-mo @PHILO-HE Can I take your time to help review this PR?

rui-mo commented on velox/core/QueryConfig.h:
```diff
@@ -304,6 +304,13 @@ class QueryConfig {
+  /// When true, establishing the result type of an arithmetic operation
+  /// happens according to Hive behavior and SQL ANSI 2011 specification, i.e.
+  /// rounding the decimal part of the result if an exact representation is not
+  /// possible. Otherwise, NULL is returned in those cases, as previously
```
> Otherwise, NULL is returned in those cases, as previously

nit: comment suggestion: NULL is returned when the actual result cannot be represented with the calculated decimal type.

rui-mo commented on velox/docs/configs.rst:
```diff
@@ -669,3 +669,8 @@ Spark-specific Configuration
+     - bool
+     - When true, establishing the result type of an arithmetic operation happens according to Hive behavior and SQL ANSI 2011 specification, i.e.
+       rounding the decimal part of the result if an exact representation is not
+       possible. Otherwise, NULL is returned in those cases, as previously
```
Ditto.

rui-mo commented on velox/functions/sparksql/DecimalUtil.h:
```diff
@@ -211,6 +211,17 @@ class DecimalUtil {
+  /// This method is used when
+  /// `spark.sql.decimalOperations.allowPrecisionLoss` is set to false.
+  /// Make sure the precision and scale is in range.
+  inline static std::pair<uint8_t, uint8_t> bounded(
```
nit: comment suggestion: Caps precision and scale at 38.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -782,21 +838,41 @@ std::shared_ptr<exec::VectorFunction> createDecimalFunction(
+    createDecimalFunction<Divide>);
+
+VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
+    udf_decimal_add_not_allow_precision_loss,
```
> This decimal function signature is only used in test, so register it in test scope.
Downstream project can only set the query config to enable this feature.

Looks like the declaration is only used by test code. Can we also test the false behavior via setting the query config?

jinchengchenghh replied:
I have added

rui-mo replied:
Do we still need the declarations and registrations then? 
https://github.com/facebookincubator/velox/pull/10383/files#diff-257bd1947469dfd8494c7e19f535d4a3247b3df086287110e53740c199104073R27-R37

jinchengchenghh replied:
Yes, it is different, with the registrations , we can test by the function name.

rui-mo replied:
I feel these declarations need to be reconsidered for below reasons.
1) they are in the runtime code but are only used by tests.
2) their signatures are generated with `allowPrecisionLoss` being false, but `createDecimalFunction`  follows the query config to decide if the function allows precision loss. Therefore, its behavior can vary according to config but its signature is fixed.

To solve above issues, maybe we need to register them as special forms. How do you think? Thanks.

jinchengchenghh replied:
For the first question, we can make the class be public, then the signature can only be in test code.
For the second issue, for all the decimal functions, we just need signature in test code, can we support to register function without function signature?

What do you think? @mbasmanova

jinchengchenghh replied:
I could change it like udf_concat_row, this will be a small refactor. 
```
VELOX_DECLARE_VECTOR_FUNCTION_WITH_METADATA(
    udf_concat_row,
    std::vector<std::shared_ptr<exec::FunctionSignature>>{},
    exec::VectorFunctionMetadataBuilder().defaultNullBehavior(false).build(),
    std::make_unique<RowFunction>());
```

jinchengchenghh replied:
And it still has a problem, we need to register the decimal add name from `add` to `decimal_add` because the name should be unique in special form, can we register it as `add` for substrait plan conversion?

jinchengchenghh replied:
We should also change here https://github.com/facebookincubator/velox/blob/main/velox/expression/ExprCompiler.cpp#L407 to break the special form name is unique, check if we can constructSpecialForm
Can we do that? @mbasmanova @rui-mo 

Which way is better?
1. current implementation;
2. special form registered as `decimal_add`, substrait function resolution unique logic;
3. special form resgistered as `add`, break special form name unique constraint.

jinchengchenghh replied:
Refactor to current version, remove the most of the signatures, only keep input type as decimal and output type as unknown, construct `Expr` to test.

jinchengchenghh replied:
Now the function signature requires `returnType`, but only `input type` and `function name` should specify the function, this also aligns with the sql function definition, we query by `select hash(col);` without return type.
And we don't permit the function with same argument and different return type, so I propose to change `returnType` in `FunctionSignature` to be optional, and for decimal functions, we don't need to supply returnType, the signature will be
```
  return {exec::FunctionSignatureBuilder()
              .integerVariable("a_precision")
              .integerVariable("a_scale")
              .integerVariable("b_precision")
              .integerVariable("b_scale")
              .argumentType("DECIMAL(a_precision, a_scale)")
              .argumentType("DECIMAL(b_precision, b_scale)")
              .build()};
```
@mbasmanova How about that?

jinchengchenghh replied:
CC @zhouyuan

rui-mo reviewed (commented):
Thanks! Added several comments.

jinchengchenghh commented:
Thanks for your comments, I have address all the comments. @rui-mo

PHILO-HE commented on velox/core/QueryConfig.h:
```diff
@@ -304,6 +304,14 @@ class QueryConfig {
+  /// When true, establishing the result type of an arithmetic operation
+  /// according to Hive behavior and SQL ANSI 2011 specification, i.e.
+  /// rounding the decimal part of the result if an exact representation is not
+  /// possible. Otherwise, NULL is returned when the actual result cannot be
```
Can you add some tests for NULL result case?

jinchengchenghh commented on velox/functions/sparksql/tests/DecimalArithmeticTest.cpp:
```diff
@@ -522,5 +545,50 @@ TEST_F(DecimalArithmeticTest, decimalDivTest) {
 }
+
+TEST_F(DecimalArithmeticTest, notAllowPrecisionLoss) {
+  setDecimalOperationsAllowPrecisionLoss(false);
```
Yes, because we construct the class Function by this config https://github.com/facebookincubator/velox/pull/10383/files#diff-4baee9f82f9c347f753972415d345941d2c2a110b608d480b090650171da096bR754

jinchengchenghh replied:
And this function name only exists in test code

mbasmanova commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -778,25 +728,40 @@ std::shared_ptr<exec::VectorFunction> createDecimalFunction(
 }
+
+// Find the function according to input types under expression compiler.
+// Mark the return type as unknown, the result type will be computed by Function
```
This is a hack. Let's not do this.

It seems like function signatures are affected by configuration settings. If that's the case, let's come up with a way to allow access to configs during function registration time.

We do need to be able to resolve function calls without instantiating functions. This is what enables Fuzzer and unit testing.

FelixYBW commented:
@jinchengchenghh The PR is used by oap/velox, can you follow up?

jinchengchenghh commented:
Address all the comments, can you help review again? @rui-mo

rui-mo commented on velox/docs/functions/spark/config.rst:
```diff
@@ -0,0 +1,21 @@
+SparkRegistration Configuration properties
+========================
+
+Generic Configuration
```
`allowPrecisionLoss` only applies to several functions so it is not a generic configuration. I feel this document is not needed as the description in `struct SparkRegistrationConfig`  is clear enough.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +27,28 @@ Multiplication
+    p = intDig + decDig
+    s = decDig
+
 For above arithmetic operators, when the precision of result exceeds 38,
```
This part applies for add/subtract/multiply/divide instead of just divide, to make this clear, could you help add a sub-title like 'Decimal Precision and Scale Adjustment'?

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +27,28 @@ Multiplication
     p = p1 - s1 + s2 + max(6, s1 + p2 + 1)
     s = max(6, s1 + p2 + 1)
 
+When config ``SparkRegistrationConfig.allowPrecisionLoss`` is set to false.
```
> If the function is registered with \``allowPrecisionLoss\`` being false:

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +27,28 @@ Multiplication
+::
+
+    intDig = min(38, p1 - s1 + s2);
+    decDig = min(38, max(6, s1 + p2 + 1));
```
nit: add one empty line below.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +27,28 @@ Multiplication
+If ``indDig + decDig`` is more than 38:
+::
+    p = 38
+    s = decDig - (intDig + decDig - 38) / 2 - 1
```
ditto.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -43,6 +57,8 @@ precision and scale are adjusted.
     precision = 38
     scale = max(38 - (p - s), min(s, 6))
 
+Caps p and s at 38 when not allowing precision loss.
```
Perhaps move below part here to illustrate how precision and scale are adjusted when allowPrecisionLoss is false.

> If ``indDig + decDig`` is more than 38:
::
    p = 38
    s = decDig - (intDig + decDig - 38) / 2 - 1

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +27,28 @@ Multiplication
 Division
 --------
-
+When config ``SparkRegistrationConfig.allowPrecisionLoss`` is set to true as default.
```
Document suggestion: 
> If the function is registered with \``allowPrecisionLoss\`` being true (default behavior):

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +27,28 @@ Multiplication
 Division
 --------
-
+When config ``SparkRegistrationConfig.allowPrecisionLoss`` is set to true as default.
```
Since the description at L5 only covers part of case now, would you add some introduction for allowPrecisionLoss being false there?

jinchengchenghh replied:
For add, subtract, multiply, the precision and scale computation is same, only adjust or bounded them.
But for divide, the precision and scale computation is different, so I only add for divide.
And document bounded behavior in the below.

rui-mo reviewed (commented):
Thanks. Added several comments on the documentation.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -1,802 +0,0 @@
-    typename A /* Argument1 */,
-    typename B /* Argument2 */,
-    typename Operation /* Arithmetic operation */>
-class DecimalBaseFunction : public exec::VectorFunction {
```
This PR also changes vector functions as simple functions. Does it make sense to do the vector-to-simple refactor in a separate PR, and support allowPrecisionLoss in this one?

jinchengchenghh replied:
Do you agree? If you think it is OK, I will only use `SparkRegistrationConfig` in this PR and revert to vector function. @mbasmanova

rui-mo replied:
Perhaps we don't need to change back to vector function in this PR. After the separate PR lands, we can do a rebase to it.

rui-mo commented on velox/functions/sparksql/DecimalFunctions.cpp:
```diff
@@ -0,0 +1,775 @@
+      return DecimalUtil::adjustPrecisionScale(precision, scale);
+    } else {
+      auto intDig = std::min(38, aPrecision - aScale + bScale);
+      auto decDig = std::min(38, std::max(6, aScale + bPrecision + 1));
```
Name suggestion: 
intDig -> wholeDigits, decDig -> fractionDigits

rui-mo commented on velox/functions/sparksql/DecimalFunctions.cpp:
```diff
@@ -0,0 +1,775 @@
+  }
+}
+
+std::vector<exec::SignatureVariable> getDivideConstraitsAllowPrecisionLoss() {
```
Constraits -> Constraints, ditto for the others.

jinchengchenghh commented:
Do you have any more comments? @rui-mo

rui-mo commented on velox/docs/functions/spark/config.rst:
```diff
@@ -0,0 +1,21 @@
+================================
+SparkRegistration Configuration
```
Title suggestion:
> Spark function registration configurations

rui-mo commented on velox/docs/functions/spark/config.rst:
```diff
@@ -0,0 +1,21 @@
+================================
+
+struct SparkRegistrationConfig
+---------------------
```
nit: make the `-` of the same length as the letters above it.

rui-mo commented on velox/docs/functions/spark/config.rst:
```diff
@@ -0,0 +1,21 @@
+     - true
+     - When true, establishing the result type of an arithmetic operation according to Hive behavior and SQL ANSI 2011 specification, i.e.
+       rounding the decimal part of the result if an exact representation is not
+       possible. Otherwise, NULL is returned when the actual result cannot be represented with the calculated decimal type. Now we support add,
```
> NULL is returned

Do we expect NULL output or an exception when the exact representation is not possible?

jinchengchenghh replied:
Yes.
AllowPrecisionLoss mode also returns null.
```
spark-sql (default)> select cast("99999999999999999999999999999999999" as decimal(38, 4)) *  cast("999999999999999999999999999999999999" as decimal(38, 4));
NULL
Time taken: 2.298 seconds, Fetched 1 row(s)
```

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -253,10 +254,12 @@ struct DecimalAddSubtractBase {
   // If result is representable with long decimal, the result
   // scale is the maximum of 'aScale' and 'bScale'. If not, reduces result scale
   // and caps the result precision at 38.
+  // Caps p and s at 38 when not allowing precision loss.
```
Do we `Caps p and s at 38` when allowPrecisionLoss being true or false?

jinchengchenghh replied:
Not allow precision and scale loss mode just `Caps p and s at 38` while another one may reduce the scale.

zhouyuan commented on velox/functions/sparksql/RegistrationConfig.h:
```diff
@@ -0,0 +1,27 @@
+  /// possible. Otherwise, NULL is returned when the actual result cannot be
+  /// represented with the calculated decimal type. Now we support add,
+  /// subtract, multiply and divide operations.
+  bool allowPrecisionLoss = true;
```
it seems try to enabled per-function config? in Spark this should be a session level config and users may modify it

rui-mo replied:
Looks like current implementation only provides fixed-behavior after the functions are registered. We might need query config if we need to make them adjustable at runtime.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -43,6 +46,26 @@ precision and scale are adjusted.
+::
+
+    wholeDigits = min(38, p1 - s1 + s2);
+    fractionalDigits = min(38, max(6, s1 + p2 + 1));
```
The `Decimal Precision and Scale Adjustment` section discusses how to handle the case when the calculated precision exceeds 38, so I assume this part and L66 belong elsewhere (L33 I think). Perhaps only keep the description for precision and scale adjustment at L57.

> For decimal addition, subtraction, multiplication, the precision and scale computation logic is same

This can be described in the corresponding location, such as L12 or L20.

rui-mo commented on velox/docs/spark_functions.rst:
```diff
@@ -4,6 +4,8 @@ Spark Functions
 
 The semantics of Spark functions match Spark 3.5 with ANSI OFF.
 
+Spark functions can be registered by :doc:`struct SparkRegistrationConfig <functions/spark/config>`.
```
Doc suggestion:
Spark functions can be registered with :doc:\`user specified configurations <functions/spark/config>\`.

jinchengchenghh commented:
As this discussion https://github.com/facebookincubator/velox/issues/10467, we cannot set it at runtime. @rui-mo

jinchengchenghh commented:
Any more comments? Thanks! @rui-mo

jinchengchenghh commented:
Soft reminder, do you have any more comment? @rui-mo

rui-mo commented:
Talked with @jinchengchenghh offline, and she will help update this PR to support session-level precision loss mode.

jinchengchenghh commented:
Updated, thanks! @rui-mo

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -9,6 +9,8 @@ https://cwiki.apache.org/confluence/download/attachments/27362075/Hive_Decimal_P
 
 https://msdn.microsoft.com/en-us/library/ms190476.aspx
 
+For decimal addition, subtraction, multiplication, the precision and scale computation logic is same.
```
Comment suggestion:

> Additionally, the computation of decimal division adapts to the allowing precision loss parameter, while the decimal addition, subtraction, and multiplication do not.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +29,26 @@ Multiplication
     p = p1 - s1 + s2 + max(6, s1 + p2 + 1)
     s = max(6, s1 + p2 + 1)
 
+When not allowing precision loss:
```
> When denying precision loss:

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +29,26 @@ Multiplication
+Decimal Precision and Scale Adjustment
+<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+
 For above arithmetic operators, when the precision of result exceeds 38,
```
> when the precision of result exceeds 38 and precision loss is allowed,

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -27,14 +29,26 @@ Multiplication
+
 For above arithmetic operators, when the precision of result exceeds 38,
-caps p at 38 and reduces the scale, in order to prevent the truncation of
+caps p at 38 and reduces the scale when allowing precision loss, in order to prevent the truncation of
```
remove ' when allowing precision loss'.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -43,6 +57,20 @@ precision and scale are adjusted.
     precision = 38
     scale = max(38 - (p - s), min(s, 6))
 
+Caps p at 38 and the scale should not be reduced when not allowing precision loss.
```
> When precision loss is denied, caps p at 38, and the scale should not be reduced.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -43,6 +57,20 @@ precision and scale are adjusted.
     scale = max(38 - (p - s), min(s, 6))
 
+Caps p at 38 and the scale should not be reduced when not allowing precision loss.
+For decimal addition, subtraction, multiplication, the adjustment logic is same.
```
> The below formula shows how the precision and scale are adjusted for decimal addition, subtraction, and multiplication.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -43,6 +57,20 @@ precision and scale are adjusted.
+
+::
+    precision = 38
+    scale = min(38, s)
```
This part does not render correctly. Perhaps one empty line after L63 is needed.

rui-mo commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -43,6 +57,20 @@ precision and scale are adjusted.
+    precision = 38
+    scale = min(38, s)
+
+But for decimal division, it is different as following:
```
> Decimal division uses a different formula, which is as follows:

rui-mo reviewed (commented):
Thanks. Added several comments on the documentation, and will take a look at the implementation later on.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -253,10 +254,12 @@ struct DecimalAddSubtractBase {
-  // operations following Hive's formulas.
+  // operations following Hive's formulas when `allowPrecisionLoss` is true.
   // If result is representable with long decimal, the result
   // scale is the maximum of 'aScale' and 'bScale'. If not, reduces result scale
```
The difference is only on whether 'adjustPrecisionScale' or 'bounded' is called, so we need to move the change on L257 perhaps before 'If not'.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -330,8 +337,10 @@ struct DecimalMultiplyFunction {
     auto [bPrecision, bScale] = getDecimalPrecisionScale(*inputTypes[1]);
-    auto [rPrecision, rScale] = DecimalUtil::adjustPrecisionScale(
-        aPrecision + bPrecision + 1, aScale + bScale);
+    auto [rPrecision, rScale] = allowPrecisionLoss
```
nit: if constexpr (allowPrecisionLoss)

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -458,65 +467,102 @@ struct DecimalDivideFunction {
       uint8_t aScale,
       uint8_t bPrecision,
       uint8_t bScale) {
```
Comment suggestion:

>   // When allowing precision loss, computes the result precision and scale
  // following Hive's formulas. If result is representable with long decimal,
  // the result scale is the maximum of 'aScale' and 'bScale'. If not, reduces
  // result scale and caps the result precision at 38. When denying precision
  // loss, calculates the number of whole digits and fraction digits. If the
  // total number of digits exceed 38, we reduce both the number of fraction
  // digits and whole digits to fit within this limit.

jinchengchenghh replied:
This is not right for divide, it is `std::max(6, aScale + bPrecision + 1)`.
This function is a private function of class in anonymous namespace, and the computation is precise by code,  do we really need to clarify the computation in comment? And for allow precision loss, the detailed comments exists in `DecimalUtil::adjustPrecisionScale(precision, scale)`, so I will comment it shortly.
> If result is representable with long decimal,
// the result scale is the maximum of 'aScale'  and 'bScale'

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -458,65 +467,102 @@ struct DecimalDivideFunction {
       ShortDecimal<P2, S2>>({name}, constraints);
 }
 
+std::string bounded(const std::string& value) {
```
Comment suggestion:
> // Used in function registration to generate the string to cap value at 38.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -458,65 +467,102 @@ struct DecimalDivideFunction {
+bool registerFunction(
+    const std::vector<std::string>& aliases = {},
+    const std::vector<exec::SignatureVariable>& constraints = {},
+    bool overwrite = true) {
```
This function seems to be the same as an existing function. Can we reuse it?

https://github.com/facebookincubator/velox/blob/e13c591f49b04ddbfc79f03335360f29ed574e3a/velox/functions/Registerer.h#L63

jinchengchenghh replied:
No, template bool is different, and `using funcClass = typename Func::template udf<exec::VectorExec>;` is different, so I add new binder to support template function with variable bool. 
```
template <template <class, bool> typename T, bool allowPrecisionLoss>
using ParameterBinder = TempWrapper<T<exec::VectorExec, allowPrecisionLoss>>;
```

jinchengchenghh replied:
I wonder if we should move it to [velox/velox/functions/Registerer.h](https://github.com/facebookincubator/velox/blob/e13c591f49b04ddbfc79f03335360f29ed574e3a/velox/functions/Registerer.h#L63) in a separate PR?
It could allow function with constraints accept template.

rui-mo replied:
It seems compared with the below function, the new one added 'constraints'. It this the only difference? Thanks.

https://github.com/facebookincubator/velox/blob/e13c591f49b04ddbfc79f03335360f29ed574e3a/velox/functions/Registerer.h#L45-L47

jinchengchenghh replied:
Yes. I wonder if we can consolidate these APIs.

rui-mo replied:
I agree. Perhaps we can change the API in a separate PR.

jinchengchenghh replied:
Do we need to only keep this one or keep both functions but the latter changes from `Func<exec::VectorExec>` to `typename Func::template udf<exec::VectorExec>`, for the backward compatibility, I would prefer keep both. The first could revoke the latter with the argument `constraints = {}`.
```
bool registerFunction(
    const std::vector<std::string>& aliases = {},
    const std::vector<exec::SignatureVariable>& constraints = {},
    bool overwrite = true) {
using funcClass = typename Func::template udf<exec::VectorExec>;
}
```
As the comment indicates, the wrapper for udf struct will be removed finally, so can I do that? @mbasmanova 
CC @rui-mo 

> // New registration function; mostly a copy from the function above, but taking
> // the inner "udf" struct directly, instead of the wrapper. We can keep both for
> // a while to maintain backwards compatibility, but the idea is to remove the
> // one above eventually.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -539,7 +585,11 @@ void registerDecimalAddSubtract(const std::string& name) {
+  registerDecimalBinary<Func, true>(
+      name, makeConstraints(rPrecision, rScale, true));
+  registerDecimalBinary<Func, false>(
+      name + "_not_allow_precision_loss",
```
The suffix '_not_allow_precision_loss' is used several times. Consider defining a variable for it and renaming as '_deny_precision_loss'.

rui-mo reviewed (commented):
Thanks.

jinchengchenghh commented:
Addressed all the comments, could you help review again? Thanks! @rui-mo

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -539,19 +577,46 @@ void registerDecimalAddSubtract(const std::string& name) {
+using AddFunctionAllowPrecisionLoss = DecimalAddFunction<TExec, true>;
+
+template <typename TExec>
+using AddFunctionDenyPrecisionLoss = DecimalAddFunction<TExec, false>;
```
The AddFunctionAllowPrecisionLoss and AddFunctionAllowPrecisionLoss are used only once. Do we need them?

jinchengchenghh replied:
Yes, because it is template function, otherwise, we need to use previous ParameterBinder.

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -561,11 +626,47 @@ void registerDecimalMultiply(const std::string& prefix) {
+      makeConstraints(rPrecision, rScale, false));
+}
+
+std::vector<exec::SignatureVariable> getDivideConstraintsDenyPrecisionLoss() {
```
Can we put them in the anonymous namespace?

rui-mo commented on velox/functions/sparksql/DecimalArithmetic.cpp:
```diff
@@ -561,11 +626,47 @@ void registerDecimalMultiply(const std::string& prefix) {
 }
 
-std::vector<exec::SignatureVariable> getDivideConstraints() {
+std::vector<exec::SignatureVariable> getDivideConstraintsAllowPrecisionLoss() {
```
Ditto.

rui-mo reviewed (commented):
Thanks. Just some nits.

rui-mo commented:
cc: @mbasmanova If you have any more comment. Thanks!

mbasmanova reviewed (commented):
@jinchengchenghh 

It is a little bit difficult to understand this change. Would it be possible to update PR description to 
- clarify how to set SparkRegistrationConfig.allowPrecisionLoss (where is SparkRegistrationConfig defined?)
- provide an example that illustrates the behavior difference between loss allowed and not?

jinchengchenghh commented:
Yes, thanks for your kindly review.
I forgot to update the PR description when the implement was changed, now it is registered as different names as we discussed in issue https://github.com/facebookincubator/velox/issues/10467#issuecomment-2339486865
Updated and added the example in PR description.
@mbasmanova

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -9,6 +9,9 @@ https://cwiki.apache.org/confluence/download/attachments/27362075/Hive_Decimal_P
 
 https://msdn.microsoft.com/en-us/library/ms190476.aspx
 
+Additionally, the computation of decimal division adapts to the allowing precision loss parameter,
```
@jinchengchenghh I'm seeing that PR description says "Each of the decimal operation functions is registered as two functions such as add_deny_precision_loss and add." But this doc suggests that only "divide" operation has 2 modes. These two seem to contradict each other. Would you clarify?

jinchengchenghh replied:
The total computation result type is divided to 2 parts, this is the first stage, compute the result type, division is different with others when not allowing precision loss, `adjust` or `bounded` the computed result type in the second stage, all operations differ with allow precision loss or not.

jinchengchenghh commented:
Do you have further comments? Thanks! @mbasmanova

mbasmanova commented:
@jinchengchenghh 

> Otherwise, NULL is returned in those cases, as previously.

What does this refer to? Isn't current before for an operation to fail? Would you provide an example where it returns NULL?

mbasmanova commented:
> When not allowing precision loss, not rounding the decimal part.

It might be helpful to clarify the tradeoff of using allowing vs. not allowing precision loss. Otherwise, one may wonder why would we ever what to allow precision loss. 

I assume some operations success if precision loss is allowed and fail if not. Would be nice to add such examples to PR description and docs.

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -9,6 +9,9 @@ https://cwiki.apache.org/confluence/download/attachments/27362075/Hive_Decimal_P
 
 https://msdn.microsoft.com/en-us/library/ms190476.aspx
 
+Additionally, the computation of decimal division adapts to the allowing precision loss parameter,
```
I still feel this sentence is confusing. My understanding is that allow-precision-loss flag applies to all operations, not just division. This sentence suggests that allow-precision-loss flag applies to division only. Am I missing something?

CC: @rui-mo

rui-mo replied:
Right. The allow-precision-loss flag applies to all operations, while there are two processes: 1) calculate the result precision and scale based on certain formula 2) adjust them if the calculated precision exceeds 38. Here I assume it refers to the first step, where only division is affected by the flag. All operations are affected by this flag in the second step as L62-L75 describes. Perhaps we could describe the two-step process clearer.

mbasmanova commented:
allow-precision-loss name might be confusing because 'precision' in 'allow-precision-loss' is not the same as 'precision' in decimal type. It might be helpful to clarify some more to avoid confusion.

jinchengchenghh commented:
> @jinchengchenghh
> 
> > Otherwise, NULL is returned in those cases, as previously.
> 
> What does this refer to? Isn't current before for an operation to fail? Would you provide an example where it returns NULL?

Yes, overflow may result in NULL. This description origins from Spark config description. I agree it is more likely return null for some edge cases when not allowing precision loss, I tried before to calculate the edge case, but I could not find an example, I will scan Spark test cases to try again.
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L3373

jinchengchenghh commented:
Added the example and update document, could you help review again? Thanks! @mbasmanova

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
-When calculating the result precision and scale of arithmetic operators,
-the formulas follow Hive which is based on the SQL standard and MS SQL:
+The result precision and scale computation of arithmetic operators contains two stages.
+In the first stage, the formulas follow Hive which is based on the SQL standard and MS SQL.
```
perhaps,

First stage computes precision and scale using formulas based on the SQL standard, MS SQL and Hive. The result may exceed maximum allowed precision of 38.

Second stage caps precision at 38 and either reduces the scale or not depending on allow-precision-loss flag.

For example, addition of decimal(38, 7) and decimal(10, 0) requires precision of 39 and scale of 7. Since precision exceeds 38 it needs to be capped. When allow-precision-loss, precision is capped at 38 and scale is reduced by 1 to 6. When allow-precision-loss, precision is capped at 38 as well, but scale is kept at 7. With allow-precision-loss all additions will succeed, but accuracy (number of digits after period) of some operations will be reduced. Without allow-precision-loss, some additions will fail.

For example,

The following addition fails when allow-precision-loss is false:

select cast('9999999999999999999999999999999.2345678' as decimal(38, 7)) + cast('1' as decimal(10, 0))

... TBD: include an error

and succeeds with allow-precision-loss is true:

... TBD: include result and highlight precision loss

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
+Decimal Precision and Scale Computation Formulas
+------------------------------------------------
+
+The SQL standard and MS SQL:
```
Do we need to refer to MS SQL? I see it is valuable to match SQL standard, but not sure if there is a lot of value in matching MS SQL and I don't believe we do that anywhere else.

jinchengchenghh replied:
I think not, Spark also does not refer to MS Sql, I will drop it.

rui-mo replied:
FYI. Below links are from Spark:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala#L413-L419.

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
 
 https://msdn.microsoft.com/en-us/library/ms190476.aspx
 
+Additionally, the computation of decimal division adapts to the allowing precision loss mode,
```
> adapts to the allowing precision loss mode,

adapts to allow-precision-loss mode

> while the decimal addition, subtraction, and multiplication do not.

This sounds like results of +, -, * are not affected by allow-precision-loss mode. I don't believe this is the case. Is it?

jinchengchenghh replied:
This is the first stage, only affect /

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
+--------------------------------------
+
+When allowing precision loss, rounding the decimal part of the result if an exact representation is not possible.
+Otherwise, NULL is returned in those cases, as previously.
```
> as previously

It is not clear what this refers to here. Perhaps, drop.

> Otherwise, NULL is returned in those cases, as previously.

Otherwise, returns NULL.

> rounding the decimal ...

rounds the decimal ...

Use active verbs for readability.

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
+
+When allowing precision loss, rounding the decimal part of the result if an exact representation is not possible.
+Otherwise, NULL is returned in those cases, as previously.
+Notice: some operations success if precision loss is allowed and fail if not.
```
success -> succeed

> and fail if not.

Do they fail or return NULL? There seems to be inconsistencies in the docs.

jinchengchenghh replied:
Depends on ANSI mode, now we only returns null.

jinchengchenghh replied:
I will change fail to `return NULL`

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
+::
+
+    select cast(0.1234567891011 as decimal(38, 18)) * cast(1234.1 as decimal(38, 18));
+    -- Result is 152.358023 when allowing precision loss, otherwise, returns NULL.
```
152.358023 if allow precision loss, NULL otherwise

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
     precision = 38
     scale = max(38 - (p - s), min(s, 6))
 
+When precision loss is denied, caps p at 38, and the scale should not be reduced.
```
denied -> not allowed

> and the scale should not be reduced.

and keeps scale as is.

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,47 +2,92 @@
+    precision = 38
+    scale = min(38, s)
+
+Decimal division uses a different formula, which is as following:
```
drop "which is as following:"

mbasmanova reviewed (commented):
@jinchengchenghh Thank you for iterating. I believe we are getting close. Some follow-up comments.

jinchengchenghh commented:
Updated the document, could you help review again? Thanks! @mbasmanova

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,49 +2,115 @@
-When calculating the result precision and scale of arithmetic operators,
-the formulas follow Hive which is based on the SQL standard and MS SQL:
+The result precision and scale computation of arithmetic operators contains two stages.
+First stage computes precision and scale using formulas based on the SQL standard, and Hive when allow-precision-loss is true.
```
> based on the SQL standard, and Hive when allow-precision-loss is true.

This is a bit confusing. SQL standard is different from Hive, no? Are you saying that we match SQL standard when allow-precision-loss is false and HiveQL when allow-precision-loss is true?

jinchengchenghh replied:
No, the SQL standard is same with Hive, we match both of them when allow-precision-loss is true.
Spark config description
```
val DECIMAL_OPERATIONS_ALLOW_PREC_LOSS =
    buildConf("spark.sql.decimalOperations.allowPrecisionLoss")
      .internal()
      .doc("When true (default), establishing the result type of an arithmetic operation " +
        "happens according to Hive behavior and SQL ANSI 2011 specification, i.e. rounding the " +
        "decimal part of the result if an exact representation is not possible. Otherwise, NULL " +
        "is returned in those cases, as previously.")
      .version("2.3.1")
      .booleanConf
      .createWithDefault(true)
```

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,49 +2,115 @@
+
+For example,
+
+The following queries keep accuracy or returns NULL with allow-precision-loss is false:
```
returns -> return
with -> when

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,49 +2,115 @@
+    select cast('1.1232154' as decimal(38, 7)) + cast('1' as decimal(10, 0)); -- 2.123215
+    select cast('9999999999999999999999999999999.2345678' as decimal(38, 7)) + cast('1' as decimal(10, 0)); -- NULL
+
+And succeed with allow-precision-loss is true:
```
> And succeed with ...

These same operations succeed when ...

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,49 +2,115 @@
+Decimal Precision and Scale Computation Formulas
+------------------------------------------------
+
+The Hive SQL behavior:
```
"Hive SQL" -> HiveQL ?

mbasmanova commented on velox/docs/functions/spark/decimal.rst:
```diff
@@ -2,49 +2,115 @@
+
+When allow-precision-loss is true, rounds the decimal part of the result if an exact representation is not possible.
+Otherwise, returns NULL.
+Notice: some operations succeed if precision loss is allowed and returns NULL if not.
```
returns -> return

mbasmanova reviewed (approved):
@jinchengchenghh Thank you for clarifying.

kevinwilfong commented:
@jinchengchenghh  could you rebase, and update the PR?

There are a lot of broken builds, but I know there was a bad merge yesterday, so it's probably due to that.

jinchengchenghh commented:
Rebased, and all the tests passed, thanks! @kevinwilfong
